mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,661 @@
+        """
+        pattern = "/".join(rule.split("/")[:-1])
+        pattern = pattern.replace(".*", ".+")
+        pattern = pattern.replace("*", ".+")
+        prefix = ".*" if pattern[0] == "/" else ".*/"
+        return prefix + pattern
+
+    def _get_entities(self, rule):
+        """ Extract the portion of a regex rule that refers to file paths.
+
+        Args:
+            rule (str): Regex rule specifying allowed file paths and entities.
+
+        Returns:
+            valid_entities (list): List of valid entities specified in the rule.
+        """
+        entities = rule.split("/")[-1]
+        entities = re.sub("[()]", "", entities).split("|")
+        valid_entities = []
+        for entity in entities:
+            if entity == "*" or self.valid_entity_check(entity):
+                valid_entities.append(entity)
+            else:
+                logger.warning("%s is not a valid entity. Skipping entity.", entity)
+        return valid_entities
+
+    @abstractmethod
+    def valid_entity_check(self, entity):
+        """ Determine if an entity type is valid.
+
+        Args:
+            entity (str): Name of entity to annotate.
+
+        Returns:
+            bool: Whether entity is valid.
+        """
+        return True
+
+    def annotate(self):
+        """ Annotate data based on configurations in the config.py file.
+        """
+        file_entities_map = self.annotate_file_entities_map
+        self._modify_queries(file_entities_map, action="annotate")
+
+    def unannotate(self):
+        """ Unannotate data based on configurations in the config.py file.
+        """
+        if not self.config["unannotate"]:
+            logger.warning("'unannotate' is None in the config. Nothing to unannotate.")
+            return
+        file_entities_map = self._get_file_entities_map(action="unannotate")
+        self._modify_queries(file_entities_map, action="unannotate")
+
+    def _modify_queries(self, file_entities_map, action):
+        """ Iterates through App files and annotates or unannotates queries.
+
+        Args:
+            file_entities_map (dict): A dictionary that maps a file paths
+                in an App to a list of entities.
+        """
+        query_factory = QueryFactory.create_query_factory(self.app_path)
+        for path in file_entities_map:
+            processed_queries = Annotator._get_processed_queries(
+                file_path=path, query_factory=query_factory
+            )
+            tqdm_desc = "Processing " + path + ": "
+            for processed_query in tqdm(processed_queries, ascii=True, desc=tqdm_desc):
+                entity_types = file_entities_map[path]
+                if action == "annotate":
+                    self._annotate_query(
+                        processed_query=processed_query, entity_types=entity_types
+                    )
+                elif action == "unannotate":
+                    Annotator._unannotate_query(
```

Comment:
Why do we do `Annotator._unannotate_query` for the unannotate usecase but `self._annotate_query` for the annotate case?

Reply:
It was to avoid pylint's "no-self-use" warning since the _unannotate_query() function does not use class attributes. Should I switch it back to self._unannotate_query() and add a comment that ignores the pylint error?

# pylint: disable=R0201